voice/src/main/kotlin/udp/AudioPacket.kt

```diff
@@ -1,43 +1,78 @@
 package dev.kord.voice.udp
 
-import dev.kord.common.annotation.KordVoice
-import dev.kord.voice.AudioFrame
-import dev.kord.voice.XSalsa20Poly1305Encoder
+import dev.kord.voice.XSalsa20Poly1305Codec
 import io.ktor.utils.io.core.*
 
-@OptIn(KordVoice::class)
-internal class AudioPacket(
-    frame: AudioFrame,
-    private val sequence: Short,
-    private val timestamp: Int,
-    private val ssrc: Int
+@OptIn(ExperimentalUnsignedTypes::class)
+sealed class AudioPacket(
+    val sequence: UShort,
+    val timestamp: UInt,
+    val ssrc: UInt,
+    val data: ByteArray
 ) {
-    private companion object {
-        private const val RTP_TYPE: Byte = 0x80.toByte()
-        private const val RTP_VERSION = 0x78.toByte()
-        private const val NONCE_LENGTH = 24
-        private const val RTP_HEADER_LENGTH = 12
-    }
-
-    private var audioData = frame.data
-
     @OptIn(ExperimentalIoApi::class)
-    private val header = BytePacketBuilder().also {
+    val header = BytePacketBuilder().also {
         it.writeByte(RTP_TYPE)
         it.writeByte(RTP_VERSION)
-        it.writeShort(sequence)
-        it.writeInt(timestamp)
-        it.writeInt(ssrc)
+        it.writeUShort(sequence)
+        it.writeUInt(timestamp)
+        it.writeUInt(ssrc)
     }.build().readBytes()
 
-    fun encrypt(key: ByteArray) {
-        val nonce = ByteArray(NONCE_LENGTH)
-        header.copyInto(nonce, 0, 0, RTP_HEADER_LENGTH)
-        audioData = XSalsa20Poly1305Encoder.encode(audioData, key, nonce)
+    class EncryptedPacket(
+        sequence: UShort,
+        timestamp: UInt,
+        ssrc: UInt,
+        encryptedData: ByteArray
+    ) : AudioPacket(sequence, timestamp, ssrc, encryptedData) {
+        fun decrypt(key: ByteArray): DecryptedPacket {
+            val nonce = ByteArray(NONCE_LENGTH)
+            header.copyInto(nonce, 0, 0, RTP_HEADER_LENGTH)
+            val decrypted = XSalsa20Poly1305Codec.decrypt(data, key, nonce) ?: error("couldn't decrypt audio data")
+            return DecryptedPacket(sequence, timestamp, ssrc, decrypted)
+        }
     }
 
     fun asByteReadPacket() = BytePacketBuilder().also {
         it.writeFully(header)
-        it.writeFully(audioData)
+        it.writeFully(data)
     }.build()
+
+    class DecryptedPacket(
+        sequence: UShort,
+        timestamp: UInt,
+        ssrc: UInt,
+        encryptedData: ByteArray
```

Comment:
should be `decryptedData` instead of `encryptedData`

Reply:
nice catch